src/ILCompiler.Compiler/src/Compiler/TypePreinit.cs

```diff
@@ -404,42 +438,78 @@ private Status TryScanMethod(MethodIL methodIL, Value[] parameters, Stack<Method
                                 return Status.Fail(methodIL.OwningMethod, opcode, "Finalizable class");
                             }
 
-                            if (((DefType)owningType).ContainsGCPointers)
+                            Value[] ctorParameters = new Value[ctorSig.Length + 1];
+                            for (int i = ctorSig.Length - 1; i >= 0; i--)
                             {
-                                // We don't want to end up with GC pointers in the frozen region
-                                // because write barriers can't handle that.
-                                return Status.Fail(methodIL.OwningMethod, opcode, "GC pointers");
+                                ctorParameters[i + 1] = stack.PopIntoLocation(GetArgType(ctor, i + 1));
                             }
 
                             Value instance;
-                            Value ctorArg0;
-                            if (owningType.IsValueType)
+                            if (owningType.IsDelegate)
                             {
-                                instance = new ValueTypeValue(owningType);
-                                ctorArg0 = ((ValueTypeValue)instance).CreateByRef();
+                                if (!(ctorParameters[2] is MethodPointerValue methodPointer))
+                                {
+                                    return Status.Fail(methodIL.OwningMethod, opcode, "Unverifiable delegate creation");
+                                }
+
+                                ForeignTypeInstance firstParameter = null;
+                                if (ctorParameters[1] != null)
+                                {
+                                    // We only have a way to refer to an allocated object if it's referenced from a static
+                                    // field of a different type. This conveniently matches delegates that the C# compiler creates
+                                    // for lambdas: this is more common than it sounds.
```

Comment:
As far as I can tell, the C# compiler creates the lamda delegates using inline lazy initialization, unfortunately.

Does this hit just for the case where the lambda is used or cache in the static constructor, or are there more cases where it hits?

Reply:
It works for these: 

https://github.com/dotnet/corert/pull/8204/files#diff-68fb5d0c14bb44619bdf2bff89aefd63R532